File: LISTAS/LISTA DOBLEMENTE ENLAZADA/LISTADOBLEMENTEENLAZADA.py
class NodoDoble:
    def __init__(self, dato):
        self.dato = dato
        self.siguiente = None
        self.anterior = None

class ListaDoblementeEnlazada:
    def __init__(self):
        self.cabeza = None
        self.cola = None
    
    def insertar_inicio(self, dato):
        nuevo = NodoDoble(dato)
        if not self.cabeza:
            self.cabeza = self.cola = nuevo
        else:
            nuevo.siguiente = self.cabeza
            self.cabeza.anterior = nuevo
            self.cabeza = nuevo
    
    def insertar_final(self, dato):
        nuevo = NodoDoble(dato)
        if not self.cabeza:
            self.cabeza = self.cola = nuevo
        else:
            nuevo.anterior = self.cola
            self.cola.siguiente = nuevo
            self.cola = nuevo
    
    def eliminar(self, dato):
        actual = self.cabeza
        while actual:
            if actual.dato == dato:
                if actual.anterior:
                    actual.anterior.siguiente = actual.siguiente
                else:
                    self.cabeza = actual.siguiente
                
                if actual.siguiente:
                    actual.siguiente.anterior = actual.anterior
                else:
                    self.cola = actual.anterior
                return
            actual = actual.siguiente
    
    def mostrar_adelante(self):
        elementos = []
        actual = self.cabeza
        while actual:
            elementos.append(actual.dato)
            actual = actual.siguiente
        return elementos
    
    def mostrar_atras(self):
        elementos = []
        actual = self.cola
        while actual:
            elementos.append(actual.dato)
            actual = actual.anterior
        return elementos

File: LISTAS/LISTA DOBLEMENTE ENLAZADA/LISTADOBLEMENTEENLAZADA.cpp
#include <iostream>
using namespace std;

class ListaDoblementeEnlazada {
private:
    struct NodoDoble {
        int dato;
        NodoDoble* siguiente;
        NodoDoble* anterior;
        NodoDoble(int d) : dato(d), siguiente(nullptr), anterior(nullptr) {}
    };
    
    NodoDoble* cabeza;
    NodoDoble* cola;
    
public:
    ListaDoblementeEnlazada() : cabeza(nullptr), cola(nullptr) {}
    
    void insertarInicio(int dato) {
        NodoDoble* nuevo = new NodoDoble(dato);
        if (!cabeza) {
            cabeza = cola = nuevo;
        } else {
            nuevo->siguiente = cabeza;
            cabeza->anterior = nuevo;
            cabeza = nuevo;
        }
    }
    
    void insertarFinal(int dato) {
        NodoDoble* nuevo = new NodoDoble(dato);
        if (!cabeza) {
            cabeza = cola = nuevo;
        } else {
            nuevo->anterior = cola;
            cola->siguiente = nuevo;
            cola = nuevo;
        }
    }
    
    void eliminar(int dato) {
        NodoDoble* actual = cabeza;
        while (actual) {
            if (actual->dato == dato) {
                if (actual->anterior) {
                    actual->anterior->siguiente = actual->siguiente;
                } else {
                    cabeza = actual->siguiente;
                }
                
                if (actual->siguiente) {
                    actual->siguiente->anterior = actual->anterior;
                } else {
                    cola = actual->anterior;
                }
                delete actual;
                return;
            }
            actual = actual->siguiente;
        }
    }
    
    void mostrarAdelante() {
        NodoDoble* actual = cabeza;
        while (actual) {
            cout << actual->dato << " ";
            actual = actual->siguiente;
        }
        cout << endl;
    }
    
    void mostrarAtras() {
        NodoDoble* actual = cola;
        while (actual) {
            cout << actual->dato << " ";
            actual = actual->anterior;
        }
        cout << endl;
    }
    
    ~ListaDoblementeEnlazada() {
        while (cabeza) {
            NodoDoble* temp = cabeza;
            cabeza = cabeza->siguiente;
            delete temp;
        }
    }
};